src/modules/feature-limit/feature-limit-migration.service.ts

```diff
@@ -0,0 +1,132 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { SequelizeUserRepository } from '../user/user.repository';
+import { SequelizeFeatureLimitsRepository } from './feature-limit.repository';
+import { AxiosError } from 'axios';
+import { User } from '../user/user.domain';
+import { PaymentsService } from 'src/externals/payments/payments.service';
+import { PLAN_FREE_TIER_ID } from './limits.enum';
+
+@Injectable()
+export class FeatureLimitsMigrationService {
+  private readonly delayBetweenUsers = 10;
+  private readonly maxRetries = 3;
+  private readonly paymentsAPIThrottlingDelay = 60000;
+
+  private planIdTierIdMap: Map<string, string>; // PlanId/tierId mapping
+
+  constructor(
+    private userRepository: SequelizeUserRepository,
+    private tiersRepository: SequelizeFeatureLimitsRepository,
+    private paymentsService: PaymentsService,
+  ) {}
+
+  async asignTiersToUsers() {
+    const limit = 20;
+    let offset = 0;
+    let processed = 0;
+    await this.loadTiers();
+
+    while (true) {
+      const users = await this.userRepository.findAllByWithPagination(
+        { tierId: null },
+        limit,
+        offset,
+      );
+
+      if (users.length === 0) {
+        break;
+      }
+
+      for (const user of users) {
+        await this.assignTier(user);
+        await this.delay(this.delayBetweenUsers); // Delay between requests to prevent Stripe Throttling
+      }
+
+      processed += users.length;
+      Logger.log(
+        `[FEATURE_LIMIT_MIGRATION]: Processed : ${processed}, current offset: ${offset} `,
+      );
+      offset += limit;
```

Comment:
If you are paginating by `tierId: null` and you are assigning tier to the current chunk of users, the offset is not needed as the next page with the WHERE clause will exclude those newly assigned tier users from the result.

In any case, the offset would be useful for failed cases in the case we want to advance no matter the situation. (`offset += failedUsers.length`)